lib/ansible/galaxy/dependency_resolution/dataclasses.py

```diff
@@ -313,6 +313,15 @@ def from_requirement_dict(cls, collection_req, art_mgr, validate_signature_optio
         # TODO: decide how to deprecate the old src API behavior
         req_source = collection_req.get('source', None)
         req_signature_sources = collection_req.get('signatures', None)
+        if frozenset(collection_req.keys()).difference(
+            frozenset({'name', 'version', 'type', 'source', 'signatures'})
+        ):
```

Comment:
Maybe store the resulting difference in a variable and mention the invalid keys in the error message. That makes it easier for users to find the bad entries.

Also you're repeating the valid keys twice, here and in the error message. How about storing them also in a variable so they are only listed in one place?